File: src/services/claudeMemory.ts
// ABOUTME: Frontend service for the Claude Code auto-memory interceptor.
// ABOUTME: Auto-provisions a SerenDB project + database + table on first run.

import { invoke } from "@tauri-apps/api/core";
import { isTauriRuntime } from "@/lib/tauri-bridge";
import { databases } from "@/services/databases";
import { settingsStore } from "@/stores/settings.store";

/**
 * Name of the SerenDB project that holds Claude Code memory. Per #1492 / #1509,
 * this is auto-created on first run if absent. The user can change which
 * project the interceptor uses via the Settings dropdowns (future).
 */
const CLAUDE_MEMORY_PROJECT_NAME = "claude-agent-prefs";

/**
 * Name of the SerenDB database inside the Claude memory project. Auto-created
 * on first run if absent.
 */
const CLAUDE_MEMORY_DATABASE_NAME = "claude_agent_prefs";

/**
 * DDL for the `claude_agent_preferences` table per #1492 spec. Idempotent
 * (uses IF NOT EXISTS) so re-running on every provision check is safe.
 *
 * NOTE: Claude memory is structured rows in this table — NOT memory SDK
 * records. This is a separate store from the user's conversational memory.
 */
const CLAUDE_MEMORY_TABLE_DDL = `
CREATE TABLE IF NOT EXISTS claude_agent_preferences (
  id SERIAL PRIMARY KEY,
  project_path TEXT NOT NULL,
  pref_key TEXT NOT NULL,
  pref_type TEXT NOT NULL,
  description TEXT,
  content TEXT NOT NULL,
  source_file TEXT,
  created_at TIMESTAMPTZ DEFAULT now(),
  updated_at TIMESTAMPTZ DEFAULT now(),
  UNIQUE(project_path, pref_key)
);
CREATE TABLE IF NOT EXISTS claude_agent_preference_audit (
  id SERIAL PRIMARY KEY,
  project_path TEXT NOT NULL,
  pref_key TEXT NOT NULL,
  action TEXT NOT NULL,
  source TEXT NOT NULL,
  created_at TIMESTAMPTZ DEFAULT now()
);
`.trim();

/**
 * How many times to poll a freshly-created database for readiness before
 * giving up. With a 2-second delay between attempts this gives ~3 minutes
 * total for the database's Postgres backend to finish provisioning.
 *
 * The original budget was 30 attempts / 60s, which was insufficient for
 * first-run cold-start provisioning on Windows (reported in #1524): a
 * user who upgrades to a version with auto-memory for the first time
 * triggers project + database + Postgres backend creation in one shot,
 * and the backend can take 60-120s to become queryable. 90 attempts
 * (180s) gives a generous margin without changing the 2s poll interval
 * (warm databases still return on the first or second attempt).
 */
const DATABASE_READY_MAX_ATTEMPTS = 90;
const DATABASE_READY_DELAY_MS = 2000;

/**
 * Error substrings that indicate the database exists in the metadata layer
 * but its Postgres backend is not yet routable by the `/query` endpoint.
 * Any of these on a `SELECT 1` means "wait and retry"; any OTHER error is
 * a real problem and should bubble up immediately.
 *
 * The `returned http <status>` markers anchor on the Rust formatter at
 * `src-tauri/src/claude_memory.rs::SerenDbSqlClient::run_sql` (which emits
 * `"SerenDB query returned HTTP {status}: {body}"`). 408 is the canonical
 * empty-body edge-timeout signature when the SerenDB SQL backend is cold;
 * 502/503/504 cover gateway-level transients during deploys, restarts, and
 * saturation. 5xx responses that carry a concrete body (e.g. 500 carrying
 * "Failed to connect to target database") still match via the connection
 * markers above. #1845.
 */
const DATABASE_NOT_READY_MARKERS = [
  "failed to connect to target database",
  "database not ready",
  "connection refused",
  "returned http 408",
  "returned http 502",
  "returned http 503",
  "returned http 504",
];

function isDatabaseNotReadyError(err: unknown): boolean {
  const msg = (err instanceof Error ? err.message : String(err)).toLowerCase();
  return DATABASE_NOT_READY_MARKERS.some((marker) => msg.includes(marker));
}

/**
 * Run a single SQL statement against SerenDB, retrying through the readiness
 * marker set above when the failure looks like a cold backend or a gateway
 * transient. The DDL leg of `ensureClaudeMemoryProvisioned` and the
 * `SELECT 1` probe in `waitForDatabaseReady` both go through this helper so
 * each leg gets the full 180-second cold-start budget — a single 408 on
 * either path used to crash the entire interceptor start (#1845).
 *
 * The query must be safe to re-run on retry. The Claude memory DDL is
 * idempotent (`CREATE TABLE IF NOT EXISTS`), and the readiness probe is
 * `SELECT 1`. Do not pass non-idempotent statements through here.
 */
async function runSqlWithReadinessRetry(
  projectId: string,
  branchId: string,
  databaseName: string,
  query: string,
  readOnly: boolean,
  maxAttempts: number = DATABASE_READY_MAX_ATTEMPTS,
  delayMs: number = DATABASE_READY_DELAY_MS,
  sleepFn: (ms: number) => Promise<void> = (ms) =>
    new Promise((r) => setTimeout(r, ms)),
): Promise<unknown> {
  for (let attempt = 1; attempt <= maxAttempts; attempt += 1) {
    try {
      const result = await databases.runSql(
        projectId,
        branchId,
        databaseName,
        query,
        readOnly,
      );
      if (attempt > 1) {
        console.info(
          `[ClaudeMemory] database "${databaseName}" served query after ${attempt} attempts`,
        );
      }
      return result;
    } catch (err) {
      if (!isDatabaseNotReadyError(err)) {
        throw err;
      }
      if (attempt === maxAttempts) {
        const totalSeconds = (maxAttempts * delayMs) / 1000;
        throw new Error(
          `SerenDB database "${databaseName}" did not become ready after ${maxAttempts} attempts (${totalSeconds}s total). ` +
            `This can happen on first run when the database backend is still provisioning. ` +
            `Last error: ${err instanceof Error ? err.message : String(err)}`,
        );
      }
      console.debug(
        `[ClaudeMemory] database "${databaseName}" not ready (attempt ${attempt}/${maxAttempts}); retrying in ${delayMs}ms`,
      );
      await sleepFn(delayMs);
    }
  }
  // Exhausted the loop without returning or throwing — unreachable with the
  // bounds above, but TypeScript can't prove it.
  throw new Error("runSqlWithReadinessRetry exhausted without resolving");
}

/**
 * Poll a database with `SELECT 1` until it responds successfully, the
 * error stops looking like a cold-start, or we hit the max attempt count.
 * Returns once the database is queryable; throws on terminal errors or
 * timeout.
 *
 * Thin wrapper around `runSqlWithReadinessRetry`. A freshly-created SerenDB
 * database is NOT immediately queryable via the `/query` endpoint — the
 * metadata write returns fast, but the Postgres backend takes seconds
 * (sometimes tens of seconds) to provision. The only safe pattern is:
 * create → poll until ready → DDL.
 */
export async function waitForDatabaseReady(
  projectId: string,
  branchId: string,
  databaseName: string,
  maxAttempts: number = DATABASE_READY_MAX_ATTEMPTS,
  delayMs: number = DATABASE_READY_DELAY_MS,
  sleepFn: (ms: number) => Promise<void> = (ms) =>
    new Promise((r) => setTimeout(r, ms)),
): Promise<void> {
  await runSqlWithReadinessRetry(
    projectId,
    branchId,
    databaseName,
    "SELECT 1",
    /* readOnly */ true,
    maxAttempts,
    delayMs,
    sleepFn,
  );
}

export interface ClaudeMemoryStatus {
  running: boolean;
  watching_root: string | null;
}

export interface ClaudeMemoryMigrationReport {
  persisted: number;
  failures: number;
  rendered: number;
  render_failures: number;
}

export interface ClaudeMemoryPreference {
  pref_key: string;
  pref_type: string;
  description: string | null;
  content: string;
  source_file: string | null;
  updated_at: string | null;
}

export interface ClaudeMemoryProjectIdentity {
  identifier: string;
  source: "git_remote" | "persisted_uuid" | "generated_uuid";
}

export interface ClaudeMemoryProvisioning {
  projectId: string;
  branchId: string;
  databaseName: string;
}

export interface InterceptSuccessEvent {
  path: string;
  name: string | null;
  memory_type: string;
  rendered_memory_md: string | null;
  render_error: string | null;
}

export interface InterceptFailureEvent {
  path: string;
  memory_type: string;
  error: string;
}

/**
 * A user-facing notice for a failed interceptor start. `status` is the parsed
 * HTTP status (for telemetry); `message` is body-free copy safe to show in a
 * dialog.
 */
export interface MemoryStartFailureNotice {
  status: number | undefined;
  message: string;
}

/**
 * Map a Claude-memory interceptor start failure to an actionable, body-free
 * user notice. Branches on the failure class — auth/permission (401/403),
 * billing (402), still-provisioning (missing key), other non-retryable client
 * errors (400/404/409/422), and retryable transients (408/429/5xx/network) —
 * so each reads differently and only the genuinely-retryable bucket promises an
 * automatic retry. The dialog never echoes raw server bodies (which can include
 * an HTTP 403 payload). The status is parsed from the `returned HTTP <status>`
 * marker that both the Rust `run_sql` formatter and the seren-db service errors
 * emit. #2497 Defect 3; #2506.
 */
export function classifyMemoryStartFailure(
  error: unknown,
): MemoryStartFailureNotice {
  const raw = error instanceof Error ? error.message : String(error);
  const statusMatch = /returned HTTP (\d{3})/i.exec(raw);
  const status = statusMatch?.[1]
    ? Number.parseInt(statusMatch[1], 10)
    : undefined;

  // Auth/permission failure — not transient. "Toggle it off and on" won't help.
  if (status === 401 || status === 403) {
    return {
      status,
      message:
        "Seren couldn't authorize memory storage for your account. This usually means your account is still finishing setup. Sign out and back in, or try again in a few minutes. If it keeps happening, contact support.",
    };
  }

  // Quota exceeded / payment required — retrying never helps; the user must top
  // up. seren-core returns 402 for this on the seren-db path. #2506.
  if (status === 402) {
    return {
      status,
      message:
        "Seren memory storage needs an active plan or balance on your account. Add funds in Settings → Wallet, then retry from Settings → Code Indexing → Claude Code Auto-Memory.",
    };
  }

  // The desktop API key hasn't landed yet — a transient provisioning failure
  // upstream kept the session but left no key. Re-authenticating re-mints it.
  if (/api key not available/i.test(raw)) {
    return {
      status,
      message:
        "Seren is still finishing your account setup, so memory storage isn't ready yet. Sign out and back in, or try again shortly.",
    };
  }

  // Other non-retryable client errors — no organization (400), no active plan
  // (404), conflict (409), unprocessable (422). These will NOT self-heal, so we
  // must not promise an automatic retry. seren-core#187 documents 400/404 here.
  if (status === 400 || status === 404 || status === 409 || status === 422) {
    return {
      status,
      message:
        "Seren couldn't set up memory storage for your account. Sign out and back in, and contact support if it keeps happening.",
    };
  }

  // Everything else (5xx, 408, 429, gateway timeout, network, cold-start
  // exhaustion) is transient and self-heals on the next memory write.
  return {
    status,
    message:
      "Seren couldn't reach memory storage just now. It will retry automatically on the next memory write — you can also retry from Settings → Code Indexing → Claude Code Auto-Memory.",
  };
}

/**
 * Resolve the SerenDB project + branch + database the Claude memory
 * interceptor should write to. Auto-creates anything missing on first run
 * and persists the resolved IDs to settings so subsequent runs reuse them.
 *
 * Idempotent: safe to call on every interceptor start. The DDL is also
 * idempotent (CREATE TABLE IF NOT EXISTS).
 */
export async function ensureClaudeMemoryProvisioned(): Promise<ClaudeMemoryProvisioning> {
  // Fast path: already provisioned and persisted to settings.
  const persistedProject = settingsStore.get("claudeMemoryProjectId");
  const persistedBranch = settingsStore.get("claudeMemoryBranchId");
  const persistedDb = settingsStore.get("claudeMemoryDatabaseName");
  if (persistedProject && persistedBranch && persistedDb) {
    return {
      projectId: persistedProject,
      branchId: persistedBranch,
      databaseName: persistedDb,
    };
  }

  // Discover or create the project. We look up by name to avoid creating
  // duplicates if the user (or a previous run) already made one.
  const allProjects = await databases.listProjects();
  let project = allProjects.find((p) => p.name === CLAUDE_MEMORY_PROJECT_NAME);
  if (!project) {
    console.info(
      `[ClaudeMemory] auto-provisioning SerenDB project "${CLAUDE_MEMORY_PROJECT_NAME}"`,
    );
    project = await databases.createProject(CLAUDE_MEMORY_PROJECT_NAME);
  }

  // Resolve the default branch (every project has a `main` branch on creation).
  const branches = await databases.listBranches(project.id);
  const branch = branches[0];
  if (!branch) {
    throw new Error(
      `SerenDB project "${CLAUDE_MEMORY_PROJECT_NAME}" has no branches; cannot provision Claude memory storage.`,
    );
  }

  // Discover or create the database within that branch.
  const allDatabases = await databases.listDatabases(project.id, branch.id);
  let database = allDatabases.find(
    (d) => d.name === CLAUDE_MEMORY_DATABASE_NAME,
  );
  if (!database) {
    console.info(
      `[ClaudeMemory] auto-provisioning SerenDB database "${CLAUDE_MEMORY_DATABASE_NAME}"`,
    );
    const created = await databases.createDatabase(
      project.id,
      branch.id,
      CLAUDE_MEMORY_DATABASE_NAME,
    );
    // Re-fetch through list so we get the full type (includes owner_name etc).
    const refreshed = await databases.listDatabases(project.id, branch.id);
    database = refreshed.find((d) => d.id === created.id) ?? undefined;
  }
  if (!database) {
    throw new Error(
      `Failed to find or create database "${CLAUDE_MEMORY_DATABASE_NAME}" in SerenDB project "${CLAUDE_MEMORY_PROJECT_NAME}".`,
    );
  }

  // Wait for the database backend to be queryable. A freshly-created
  // SerenDB database is NOT immediately reachable via the `/query`
  // endpoint — the metadata write returns fast, but the Postgres backend
  // takes seconds to provision. Poll SELECT 1 until it responds.
  console.info(
    `[ClaudeMemory] waiting for database "${CLAUDE_MEMORY_DATABASE_NAME}" to become ready`,
  );
  await waitForDatabaseReady(
    project.id,
    branch.id,
    CLAUDE_MEMORY_DATABASE_NAME,
  );

  // Apply the table DDL through the same readiness-retry helper as the
  // SELECT 1 probe. The DDL is idempotent (CREATE TABLE IF NOT EXISTS), and
  // a cold backend or transient gateway 408/502/503/504 between the probe
  // and the DDL must not collapse the whole start (#1845).
  console.info(
    `[ClaudeMemory] applying claude_agent_preferences DDL to ${CLAUDE_MEMORY_DATABASE_NAME}`,
  );
  await runSqlWithReadinessRetry(
    project.id,
    branch.id,
    CLAUDE_MEMORY_DATABASE_NAME,
    CLAUDE_MEMORY_TABLE_DDL,
    /* readOnly */ false,
  );

  // Persist the resolved IDs so subsequent runs hit the fast path.
  settingsStore.set("claudeMemoryProjectId", project.id);
  settingsStore.set("claudeMemoryBranchId", branch.id);
  settingsStore.set("claudeMemoryDatabaseName", CLAUDE_MEMORY_DATABASE_NAME);

  return {
    projectId: project.id,
    branchId: branch.id,
    databaseName: CLAUDE_MEMORY_DATABASE_NAME,
  };
}

/**
 * Start the filesystem watcher.
 *
 * Calls `ensureClaudeMemoryProvisioned()` first so the SerenDB project,
 * database, and table all exist before the watcher starts intercepting files.
 * Then passes the resolved provisioning identifiers to Rust so the watcher's
 * SQL writes know exactly where to land.
 */
export async function startClaudeMemoryInterceptor(): Promise<ClaudeMemoryStatus> {
  if (!isTauriRuntime()) {
    return { running: false, watching_root: null };
  }
  const provisioning = await ensureClaudeMemoryProvisioned();
  return invoke<ClaudeMemoryStatus>("claude_memory_start", {
    projectId: provisioning.projectId,
    branchId: provisioning.branchId,
    databaseName: provisioning.databaseName,
  });
}

/**
 * Stop the watcher. Safe to call when it is not running.
 */
export async function stopClaudeMemoryInterceptor(): Promise<ClaudeMemoryStatus> {
  if (!isTauriRuntime()) {
    return { running: false, watching_root: null };
  }
  return invoke<ClaudeMemoryStatus>("claude_memory_stop");
}

/**
 * Snapshot the watcher's current running state without mutating it.
 */
export async function getClaudeMemoryStatus(): Promise<ClaudeMemoryStatus> {
  if (!isTauriRuntime()) {
    return { running: false, watching_root: null };
  }
  return invoke<ClaudeMemoryStatus>("claude_memory_status");
}

/**
 * Walk every existing Claude memory directory and push any pre-existing `.md`
 * files to SerenDB. Returns persisted, cloud-write failures, refreshed
 * MEMORY.md indexes, and refresh failures. Files whose cloud write fails are
 * left on disk so the live watcher can retry later.
 *
 * Provisions the SerenDB project/database/table first if needed.
 */
export async function migrateExistingClaudeMemory(): Promise<ClaudeMemoryMigrationReport> {
  if (!isTauriRuntime()) {
    return { persisted: 0, failures: 0, rendered: 0, render_failures: 0 };
  }
  const provisioning = await ensureClaudeMemoryProvisioned();
  return invoke<ClaudeMemoryMigrationReport>("claude_memory_migrate_existing", {
    projectId: provisioning.projectId,
    branchId: provisioning.branchId,
    databaseName: provisioning.databaseName,
  });
}

/**
 * Resolve a stable identifier for a project directory — git remote URL or a
 * persisted UUID at `<cwd>/.claude/project_id`.
 */
export async function getClaudeProjectIdentity(
  projectCwd: string,
): Promise<ClaudeMemoryProjectIdentity | null> {
  if (!isTauriRuntime()) {
    return null;
  }
  return invoke<ClaudeMemoryProjectIdentity>(
    "claude_memory_get_project_identity",
    { projectCwd },
  );
}

/**
 * Returns true when Claude CLI's session JSONL file exists on disk for the
 * given project cwd + session id. Used by `resumeAgentConversation` to skip
 * `--resume` when the stored session ID points at a missing file (CLI
 * cleaned up old sessions, app reinstall, cross-machine sync) — without this
 * pre-flight, the spawn fails with `code=1: No conversation found with
 * session ID: <id>` and surfaces a "Claude Code request failed" error event.
 * Browser runtime: returns false (no CLI sessions to check). See #1657.
 */
export async function claudeSessionExists(
  projectCwd: string,
  sessionId: string,
): Promise<boolean> {
  if (!isTauriRuntime()) {
    return false;
  }
  return invoke<boolean>("claude_session_exists", { projectCwd, sessionId });
}

/**
 * Render `~/.claude/projects/<encoded(projectCwd)>/memory/MEMORY.md` from the
 * `claude_agent_preferences` SerenDB table, so Claude Code reads fresh
 * content at the start of its next session.
 */
export async function renderClaudeMemoryMd(
  projectCwd: string,
): Promise<string | null> {
  if (!isTauriRuntime()) {
    return null;
  }
  const provisioning = await ensureClaudeMemoryProvisioned();
  return invoke<string>("claude_memory_render_memory_md", {
    projectCwd,
    projectId: provisioning.projectId,
    branchId: provisioning.branchId,
    databaseName: provisioning.databaseName,
  });
}

/**
 * Recall one Claude Code auto-memory body from SerenDB by `pref_key`.
 * This is the supported read path for bodies listed in MEMORY.md; the
 * original sibling `.md` files are removed after a successful cloud write.
 */
export async function recallClaudeMemoryPreference(
  projectCwd: string,
  prefKey: string,
): Promise<ClaudeMemoryPreference | null> {
  if (!isTauriRuntime()) {
    return null;
  }
  const provisioning = await ensureClaudeMemoryProvisioned();
  return invoke<ClaudeMemoryPreference | null>(
    "claude_memory_recall_preference",
    {
      projectCwd,
      prefKey,
      projectId: provisioning.projectId,
      branchId: provisioning.branchId,
      databaseName: provisioning.databaseName,
    },
  );
}

File: src-tauri/src/commands/claude_memory.rs
// ABOUTME: Tauri commands for the Claude Code auto-memory interceptor.
// ABOUTME: Exposes start/stop/status, startup migration, render, and identity lookup.

use std::path::{Path, PathBuf};
use std::time::Duration;

use serde::Serialize;
use tauri::AppHandle;

use crate::claude_memory::{self, QueryResult, SerenDbConfig};

/// Hard upper bound on how long `claude_memory_start` is allowed to take
/// before returning a timeout error to the caller. The watcher does an
/// `fs::create_dir_all`, an FSEvents recursive watch on `~/.claude/projects`,
/// and a `tokio::spawn` — none of which should take more than a fraction of
/// a second on a healthy system, but we cap it so a slow filesystem cannot
/// freeze the UI through the IPC dispatch.
const CLAUDE_MEMORY_START_TIMEOUT: Duration = Duration::from_secs(10);

/// Public status snapshot for the frontend.
#[derive(Debug, Serialize)]
pub struct ClaudeMemoryStatus {
    pub running: bool,
    pub watching_root: Option<String>,
}

/// Public project identity payload for the frontend.
#[derive(Debug, Serialize)]
pub struct ClaudeMemoryProjectIdentity {
    pub identifier: String,
    pub source: String,
}

fn build_config(
    project_id: String,
    branch_id: String,
    database_name: String,
) -> Result<SerenDbConfig, String> {
    if project_id.is_empty() || branch_id.is_empty() || database_name.is_empty() {
        return Err(
            "claude_memory_start requires non-empty projectId, branchId, and databaseName \
             (the frontend's ensureClaudeMemoryProvisioned helper should have populated these)"
                .to_string(),
        );
    }
    Ok(SerenDbConfig {
        project_id,
        branch_id,
        database_name,
    })
}

/// Start watching `~/.claude/projects/*/memory/` for Claude Code memory writes.
///
/// `claude_memory::start_watcher` does synchronous filesystem and notify-watcher
/// setup work that previously ran on the main Tauri thread. Under load (deep
/// `~/.claude/projects` tree, slow disk, contention with the auth store) that
/// could freeze the UI. We dispatch the entire body onto the blocking pool and
/// cap it with a timeout so the IPC caller is never parked indefinitely.
///
/// The frontend's `ensureClaudeMemoryProvisioned()` helper auto-creates the
/// `claude-agent-prefs` SerenDB project, the `claude_agent_prefs` database,
/// and the `claude_agent_preferences` table on first run, then passes the
/// resolved identifiers down to this command. Per #1509, Claude memory is
/// stored in a separate SerenDB project from the user's conversational
/// memory — these are two distinct stores and must not share storage.
#[tauri::command]
pub async fn claude_memory_start(
    app: AppHandle,
    project_id: String,
    branch_id: String,
    database_name: String,
) -> Result<ClaudeMemoryStatus, String> {
    let config = build_config(project_id, branch_id, database_name)?;

    let join =
        tokio::task::spawn_blocking(move || claude_memory::start_watcher(app, config));
    let root = match tokio::time::timeout(CLAUDE_MEMORY_START_TIMEOUT, join).await {
        Ok(Ok(Ok(path))) => path,
        Ok(Ok(Err(e))) => return Err(e),
        Ok(Err(join_err)) => {
            return Err(format!(
                "Claude memory interceptor start task panicked: {join_err}"
            ));
        }
        Err(_elapsed) => {
            return Err(format!(
                "Claude memory interceptor start timed out after {}s — check that ~/.claude/projects is reachable and that you are logged in to SerenDB",
                CLAUDE_MEMORY_START_TIMEOUT.as_secs()
            ));
        }
    };

    Ok(ClaudeMemoryStatus {
        running: true,
        watching_root: Some(root.to_string_lossy().to_string()),
    })
}

/// Stop the watcher if running.
///
/// Dispatched onto the blocking pool for the same reason as `claude_memory_start`:
/// the underlying `stop_watcher` acquires a global `Mutex` and calls
/// `task.abort()` — both fast in the happy path, but we never want either to
/// park the main Tauri thread.
#[tauri::command]
pub async fn claude_memory_stop() -> Result<ClaudeMemoryStatus, String> {
    let join = tokio::task::spawn_blocking(claude_memory::stop_watcher);
    match join.await {
        Ok(Ok(())) => {}
        Ok(Err(e)) => return Err(e),
        Err(join_err) => {
            return Err(format!(
                "Claude memory interceptor stop task panicked: {join_err}"
            ));
        }
    }
    Ok(ClaudeMemoryStatus {
        running: false,
        watching_root: None,
    })
}

/// Return the current watcher status without mutating it.
#[tauri::command]
pub fn claude_memory_status() -> Result<ClaudeMemoryStatus, String> {
    let running = claude_memory::is_watcher_running();
    let watching_root = if running {
        claude_memory::claude_projects_root()
            .ok()
            .map(|p| p.to_string_lossy().to_string())
    } else {
        None
    };
    Ok(ClaudeMemoryStatus {
        running,
        watching_root,
    })
}

/// Walk every existing Claude memory directory and INSERT any pre-existing
/// `.md` files into `claude_agent_preferences`. Returns persisted + failures.
#[tauri::command]
pub async fn claude_memory_migrate_existing(
    app: AppHandle,
    project_id: String,
    branch_id: String,
    database_name: String,
) -> Result<claude_memory::MigrationReport, String> {
    let config = build_config(project_id, branch_id, database_name)?;
    claude_memory::migrate_existing_files(&app, &config).await
}

/// Resolve a stable project identifier for `project_cwd` (git remote or UUID).
/// Check whether Claude CLI's session JSONL file exists on disk for the given
/// project cwd + session id. Used by the frontend to skip `--resume` when the
/// stored session ID points at a missing file (CLI cleaned up its sessions
/// dir, app reinstall, cross-machine sync). Without this pre-flight, the
/// resume attempt fails with `code=1: No conversation found with session ID:
/// <id>` and surfaces a "Claude Code request failed" error event before the
/// recovery fallback kicks in (#1657).
#[tauri::command]
pub fn claude_session_exists(
    project_cwd: String,
    session_id: String,
) -> Result<bool, String> {
    let root = claude_memory::claude_projects_root()?;
    Ok(claude_memory::session_jsonl_path(&root, Path::new(&project_cwd), &session_id).is_file())
}

#[tauri::command]
pub fn claude_memory_get_project_identity(
    project_cwd: String,
) -> Result<ClaudeMemoryProjectIdentity, String> {
    let identity = claude_memory::resolve_project_identity(Path::new(&project_cwd))?;
    let source = match identity.source {
        claude_memory::ProjectIdentitySource::GitRemote => "git_remote",
        claude_memory::ProjectIdentitySource::PersistedUuid => "persisted_uuid",
        claude_memory::ProjectIdentitySource::GeneratedUuid => "generated_uuid",
    }
    .to_string();
    Ok(ClaudeMemoryProjectIdentity {
        identifier: identity.identifier,
        source,
    })
}

/// Render `~/.claude/projects/<encoded(project_cwd)>/MEMORY.md` from the
/// `claude_agent_preferences` table in SerenDB. The frontend calls this when
/// a project is opened so Claude Code reads fresh content next session.
#[tauri::command]
pub async fn claude_memory_render_memory_md(
    app: AppHandle,
    project_cwd: String,
    project_id: String,
    branch_id: String,
    database_name: String,
) -> Result<String, String> {
    let config = build_config(project_id, branch_id, database_name)?;
    let root = claude_memory::claude_projects_root()?;
    let encoded = claude_memory::encode_project_dir(Path::new(&project_cwd));
    let claude_project_dir: PathBuf = root.join(&encoded);

    let api_key = {
        use tauri_plugin_store::StoreExt;
        app.store("auth.json")
            .map_err(|e| e.to_string())?
            .get("seren_api_key")
            .and_then(|v| v.as_str().map(|s| s.to_string()))
            .unwrap_or_default()
    };
    if api_key.is_empty() {
        return Err(
            "SerenDB API key not available — log in to Seren Desktop so the key is provisioned"
                .to_string(),
        );
    }

    let client = claude_memory::SerenDbSqlClient::new(api_key);
    let final_path =
        claude_memory::render_memory_md_from_db(&client, &config, &claude_project_dir).await?;
    Ok(final_path.to_string_lossy().to_string())
}

/// Execute a SQL statement against a SerenDB database from the frontend via
/// the Rust-side `SerenDbSqlClient` (reqwest + SerenDB API key).
///
/// This command exists because the TypeScript `databases.runSql` path has
/// failed through every other routing we tried:
///
///   1. The hey-api generated client went through the `customFetch` →
///      Rust gateway bridge, which injects the **OAuth bearer token**. The
///      SerenDB `/publishers/seren-db/query` endpoint authenticates with the
///      **SerenDB API key**, so this path returned HTTP 500 "Failed to
///      connect to target database" (#1511 diagnosis).
///
///   2. A direct cross-origin `fetch()` from the Tauri webview to
///      `api.serendb.com` was blocked by webview CORS with
///      `TypeError: Load failed` (#1512 diagnosis).
///
///   3. Routing through `callSerenTool("run_sql", ...)` via the Seren MCP
///      gateway at `mcp.serendb.com` worked architecturally (Rust-side
///      `rmcp` client has no CORS, handles its own auth) but the MCP
///      gateway fails to connect in the user's dev environment — the
///      `waitForGatewayReady` guard (#1513) just makes the failure
///      explicit after 30 seconds instead of immediately.
///
/// So we do what the Rust watcher already does successfully: call
/// `reqwest::Client` from Rust, which has no CORS, with the SerenDB API
/// key read from the Tauri secure store. This bypasses both the browser
/// origin policy AND the broken MCP gateway, and reuses the same client
/// code path that `process_memory_file` runs on every intercepted write.
///
/// The only caller is `databases.runSql` in the frontend, which is itself
/// used by the Claude memory auto-provisioning flow (`ensureClaudeMemoryProvisioned`,
/// `waitForDatabaseReady`) — i.e. the setup that runs BEFORE the watcher
/// starts. Once the watcher is running, SQL writes happen entirely in Rust
/// and never touch this command.
#[tauri::command]
pub async fn claude_memory_run_sql(
    app: AppHandle,
    project_id: String,
    branch_id: Option<String>,
    database_name: Option<String>,
    query: String,
    read_only: bool,
) -> Result<QueryResult, String> {
    if project_id.is_empty() {
        return Err("claude_memory_run_sql requires a non-empty projectId".to_string());
    }
    if query.is_empty() {
        return Err("claude_memory_run_sql requires a non-empty query".to_string());
    }

    let client = claude_memory::build_sql_client(&app)?;
    let config = SerenDbConfig {
        project_id,
        branch_id: branch_id.unwrap_or_default(),
        database_name: database_name.unwrap_or_default(),
    };
    client.run_sql(&config, &query, read_only).await
}

// ============================================================================
// Tests for #1507 (claude_memory_start must not block the main Tauri thread).
// The pattern itself is verified here; the SQL builder + injection-prevention
// tests live in `claude_memory::tests` since that's where the SQL helpers live.
// ============================================================================

#[cfg(test)]
mod tests {
    use super::*;
    use std::time::Instant;

    #[tokio::test(flavor = "multi_thread")]
    async fn start_pattern_returns_within_bound_when_inner_blocks() {
        // The inner closure simulates a `start_watcher` call that has parked
        // on a slow filesystem, an FSEvents init, or a contended Mutex —
        // exactly the failure mode #1507 reported. The wrapping
        // spawn_blocking + timeout MUST return control to the caller within
        // the timeout bound, otherwise the production fix is not actually
        // protecting the main Tauri thread.
        let inner_blocking_duration = Duration::from_millis(1500);
        let outer_timeout = Duration::from_millis(500);
        let started = Instant::now();

        let join = tokio::task::spawn_blocking(move || {
            std::thread::sleep(inner_blocking_duration);
        });
        let result = tokio::time::timeout(outer_timeout, join).await;
        let elapsed_to_timeout = started.elapsed();

        assert!(
            result.is_err(),
            "spawn_blocking + timeout MUST return a timeout error when the inner work blocks; got {result:?}"
        );
        assert!(
            elapsed_to_timeout < Duration::from_secs(2),
            "expected timeout to fire within 2s (the main Tauri thread is not parked); took {elapsed_to_timeout:?}"
        );
    }

    #[test]
    fn claude_memory_start_timeout_constant_is_bounded() {
        assert!(
            CLAUDE_MEMORY_START_TIMEOUT <= Duration::from_secs(60),
            "CLAUDE_MEMORY_START_TIMEOUT must stay bounded; got {:?}",
            CLAUDE_MEMORY_START_TIMEOUT
        );
        assert!(
            CLAUDE_MEMORY_START_TIMEOUT >= Duration::from_secs(5),
            "CLAUDE_MEMORY_START_TIMEOUT too aggressive; got {:?}",
            CLAUDE_MEMORY_START_TIMEOUT
        );
    }

    #[test]
    fn build_config_rejects_empty_strings() {
        assert!(build_config("".into(), "b".into(), "d".into()).is_err());
        assert!(build_config("p".into(), "".into(), "d".into()).is_err());
        assert!(build_config("p".into(), "b".into(), "".into()).is_err());
        assert!(build_config("p".into(), "b".into(), "d".into()).is_ok());
    }
}
